### PR TITLE
Use Webpack & modernize Dropkick

### PR DESCRIPTION
**Pull request by Robdel12**
## What is this?

I hated the gulp setup I made 3 years ago so I decided to use webpack and modernize this repo like it badly needs. 

### What changed?

- Use webpack to compile ES6 files
- Changed dropkick to an ES6 class
- Rewrote tests to use karma & mocha
- Cross browser testing with browser stack
- `/lib` changed to `/src`
- `/build` changed to `/dist`
- Use circleci 2.0 for testing
- Use yarn for dependencies
- Use at least node 4
- Improved test coverage
- Changed the docs generation to `documentation.js`

## TODOs
- [x] Finish porting tests
- [x] Update documentation
- [x] Update README & Contributing
- [x] Switch to circle for testing
- [x] Fill out PR description with all the different changes

### CODE REVIEW

File: tests/dropkick-test.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
+import { expect } from 'chai';
+import { $ } from './utils.js';
+import Dropkick from '../src/dropkick.js';
+
+
+function buildSelect(id, options) {
```
Comment (Robdel12) [inline tests/dropkick-test.js]: Move this to test utils

Comment (Robdel12) [inline tests/dropkick-test.js]: Also do an after each clean up of the selects that were built

File: tests/mobile-test.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+// });
+
+// QUnit.test( "Dropkick shouldn't blow up when calling methods on mobile: false dks", 1, function( assert ) {
+//   var $sel = $("#normal_select", {mobile: false}),
+//       dk = new Dropkick($sel[0]);
+
+//   assert.equal(navigator.userAgent, "Android");
+
+//   dk.refresh();
+// });
```
Comment (Robdel12) [inline tests/mobile-test.js]: This can probably go away and use browser stack to run our tests on mobile devices

File: tests/iframe-test.js [hunk 3]
```diff
@@ -1,9 +1,9 @@
...
-QUnit.test( "Dropkick closes from click event in parent document when running inside an iframe", function( assert ) {
+// QUnit.test( "Dropkick closes from click event in parent document when running inside an iframe", function( assert ) {
 //   var dk = new Dropkick("#normal_select");
 
 //   dk.open();
 //   assert.equal(dk.isOpen, true);
 //   $(".body").trigger("click");
 //   assert.equal(dk.isOpen, false);
-});
+// });
```
Comment (Robdel12) [inline tests/iframe-test.js]: This probably can get moved into the main test file and create a helper that builds / cleans up iframe creation

File: tests/src/runner.html [hunk 4]
```diff
@@ -119,12 +119,7 @@
...
-  <!-- <script src="../../lib/polyfills/customevent.js"></script>
-  <script src="../../lib/polyfills/eventtarget.js"></script>
-  <script src="../../lib/polyfills/foreach.js"></script>
-  <script src="../../lib/polyfills/indexof.js"></script>
-  <script src="../../lib/dropkick.js"></script> -->
-  <script src="../../build/js/dropkick.min.js"></script>
+  <script src="../../dist/dropkick.js"></script>
   <script src="jquery-1.11.2.js"></script>
   <script src="../../node_modules/qunitjs/qunit/qunit.js"></script>
   <script src="../unit/tests.js"></script>
```
Comment (Robdel12) [inline tests/src/runner.html]: Delete this

File: dist/index.html [hunk 5]
```diff
@@ -1,71 +1,72 @@
...
+      <option value="WV">West Virginia</option>
+      <option value="WI">Wisconsin</option>
+      <option value="WY">Wyoming</option>
+    </select>
+    <button onclick="close()">Close</button>
+    <script src="dropkick.js"></script>
+    <script>
+     var select = document.getElementById('normal_select');
+     var dropkick = new Dropkick(select);
+     console.log(dropkick);
```
Comment (wwilsman) [inline dist/index.html]: 🤔

File: dist/index.html [hunk 6]
```diff
@@ -1,71 +1,72 @@
...
+     function close() {
```
Comment (wwilsman) [inline dist/index.html]: Will `dropkick.close()` not work in the `onclick` up above?

File: karma.conf.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      require('karma-chrome-launcher')
+    ]
+  });
+
+  // CI config
+  if (process.env.TRAVIS || process.env.CI) {
+    config.singleRun = true;
+    config.browsers = ['Chrome_travis_ci'];
+  }
+};
```
Comment (wwilsman) [inline karma.conf.js]: [A lot of these options are defaults that we could safely remove.](http://karma-runner.github.io/1.0/config/configuration-file.html)

File: tests/mobile-test.js (hunk 2 above)
Comment (wwilsman) [inline tests/mobile-test.js]: Should add this comment to the file so it doesn't get lost. :wink:

File: webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+const isProduction = process.env.NODE_ENV === "production";
+const UglifyJSPlugin = require('uglifyjs-webpack-plugin');
+const ExtractTextPlugin = require("extract-text-webpack-plugin");
+const extractSass = new ExtractTextPlugin({
+  filename: "dropkick.css"
+});
+
+// I hate this, but I guess you can't pass disable to uglify..
+const plugins = [ extractSass ];
+isProduction ? plugins.push(new UglifyJSPlugin()) : null ;
```
Comment (wwilsman) [inline webpack.config.js]: If you don't like this, what about [this](https://github.com/wwilsman/monopoly-wallet/blob/master/webpack.config.js)?

I thought it was tidy and reusable. Food for thought ¯\\\_(ツ)\_/¯

File: dist/index.html (hunk 6 above)
Comment (Robdel12) [inline dist/index.html]: Yea, this was me hacking around before the test suite was up

File: tests/mobile-test.js (hunk 2 above)
Comment (Robdel12) [inline tests/mobile-test.js]: I reached out to browser stack to see if we can get an account to run tests. I’ve got the setup going locally but I’m about to run out of test free test runs

File: karma.conf.js (hunk 7 above)
Comment (Robdel12) [inline karma.conf.js]: - [ ] TODO

File: webpack.config.js (hunk 8 above)
Comment (Robdel12) [inline webpack.config.js]: It felt more dirty than being able to do: 
```js
new UglifyJsPlugin({
  disable: !isProduction
})
```

But I guess this isn’t terrible.

File: karma.conf.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      require('karma-mocha-reporter'),
+      require('karma-browserstack-launcher'),
+      require('karma-chrome-launcher')
+    ]
+  });
+
+  // CI config
+  if (process.env.TRAVIS || process.env.CI) {
+    config.singleRun = true;
+    config.browsers = ['Chrome_travis_ci'];
```
Comment (Robdel12) [inline karma.conf.js]: This should include the browser stack launchers too

**Comment by Robdel12** [comment]
@wwilsman wanna give one last go?

File: webpack.config.js (hunk 8 above)
Comment (wwilsman) [inline webpack.config.js]: Also I realize the ternary is moot. Just do

```
isProduction && plugins.push(...);
```

Or explicitly
```
if (isProduction) { plugins.push(...); }
```

**Comment by wwilsman** [comment]
🎉

**Pull request merged**